### PR TITLE
EpiModel: Get/Set Functionality

### PR DESCRIPTION
**Pull request by andsv2**
Implement functionality of new get/set helper functions into EpiModel base modules.

https://github.com/statnet/EpiModel/issues/374

### CODE REVIEW

**Comment by andsv2** [comment]
Afternoon, @smjenness. We've got the get/set functionality passing on Travis. Had to make some changes to internal functions to get this going. Most "significant" was re-ordering of arrivals/departures cleanup in nwupdate: this has to go first so that network object and attributes are consistent.

File: R/net.mod.infection.R [hunk 1]
```diff
@@ -92,8 +92,12 @@ infection.net <- function(dat, at) {
 
       # Set new infections vector
       idsNewInf <- unique(del$sus)
-      dat$attr$status[idsNewInf] <- "i"
-      dat$attr$infTime[idsNewInf] <- at
+      status.temp <- get_attr(dat, "status")
+      status.temp[idsNewInf] <- "i"
```
Comment (smjenness) [inline R/net.mod.infection.R:96]: This seems unnecessary. status and infTime can be pulled at the top of the function, and then set at the bottom of the function. Each module should follow this convention:

1. Pull from dat (param, attr, etc., needed in processes)
2. Work on vectors directly
3. Put back on dat

File: R/net.mod.nwupdate.R [hunk 2]
```diff
@@ -13,86 +13,94 @@
...
+  exitTime <- get_attr(dat, "exitTime")
+
+  ## Vital Dynamics
+
+  if (get_param(dat, "vital")) {
+
+    ## Arrivals
+    if (groups == 1) {
+      nArrivals <- get_epi(dat, "a.flow")[at]
+    } else {
```
[the remark this answers is not in the document]
Comment (smjenness) [reply R/net.mod.nwupdate.R:30]: Prefer that to using bracket notation

File: R/net.mod.simnet.R [hunk 3]
```diff
@@ -125,21 +130,32 @@ resim_nets <- function(dat, at) {
...
-      group <- dat$attr$group
-      old.num.g1 <- dat$epi$num[at - 1]
-      old.num.g2 <- dat$epi$num.g2[at - 1]
-      new.num.g1 <- sum(dat$attr$active == 1 & group == 1)
-      new.num.g2 <- sum(dat$attr$active == 1 & group == 2)
+    if (groups == 2) {
+      index <- at-1
+      if (tergmLite == FALSE){
+        group <- idgroup(dat$nw)
+      } else {
```
Comment (smjenness) [inline R/net.mod.simnet.R:152]: no need to use idgroup here even if tergmLite is FALSE. can always just lookup with get_attr. Right?

File: R/net.mod.vital.R [hunk 4]
```diff
@@ -17,68 +17,75 @@ departures.net <- function(dat, at) {
...
 
   # Susceptible departures ------------------------------------------------------
   nDepartures.sus <- 0
-  idsElig.sus <- which(dat$attr$active == 1 & dat$attr$status == "s")
+  idsElig.sus <- which(active == 1 & status == "s")
   nElig.sus <- length(idsElig.sus)
   if (nElig.sus > 0) {
-    rates.sus <- dat$param$ds.rate
+    rates.sus <- get_param(dat, "ds.rate")
     vecDepartures.sus <- which(rbinom(nElig.sus, 1, rates.sus) == 1)
```
Comment (smjenness) [inline R/net.mod.vital.R:32]: pull the rates at the top

File: R/net.mod.vital.R [hunk 5]
```diff
@@ -17,68 +17,75 @@ departures.net <- function(dat, at) {
...
   # Output ------------------------------------------------------------------
 
+  dat <- set_attr(dat, "active", active)
+  dat <- set_attr(dat, "exitTime", exitTime)
+
   if (at == 2) {
-    dat$epi$ds.flow <- c(0, nDepartures.sus)
-    dat$epi$di.flow <- c(0, nDepartures.inf)
+    dat <- set_epi(dat, "ds.flow", c(0, nDepartures.sus))
+    dat <- set_epi(dat, "di.flow", c(0, nDepartures.inf))
```
Comment (smjenness) [inline R/net.mod.vital.R:80]: similar comment to above

File: R/net.utils.R [hunk 6]
```diff
@@ -717,6 +717,10 @@ get_network_term_attr <- function(nw) {
 #' idgroup(nw, ids = c(3, 6))
 #'
 idgroup <- function(nw, ids) {
+
+  if (class(nw) == "list") {
+    nw <- nw[[1]]
```
Comment (smjenness) [inline R/net.utils.R:722]: why is this necessary? the nw argument is defined as a network object. So it would be expected that if it were embedded in a sublist, one would pass in idgroup(nw[[1]], ...)

File: R/net.mod.infection.R [hunk 7]
```diff
@@ -115,9 +119,9 @@ infection.net <- function(dat, at) {
 
   ## Save incidence vector
   if (at == 2) {
-    dat$epi$si.flow <- c(0, nInf)
+    dat <- set_epi(dat, "si.flow", c(0, nInf))
   } else {
-    dat$epi$si.flow[at] <- nInf
+    dat <- set_epi_at(dat, "si.flow", at, nInf)
   }
```
[the remark this answers is not in the document]
Comment (smjenness) [reply R/net.mod.infection.R:125]: actually, after reviewing the functionality, I see that these are two functions. I will ask @AdrienLeGuillou to update

File: R/net.utils.R (hunk 6 above)
Comment (andsv2) [reply R/net.utils.R:722]: @smjenness this was put in to differentiate between cases used outside of EpiModel and within the EpiModel workflow. This was a change mainly to accommodate the examples provided with this function.

File: R/net.mod.infection.R [hunk 8]
```diff
@@ -300,9 +312,11 @@ infection.2g.net <- function(dat, at) {
 #'
 discord_edgelist <- function(dat, at) {
 
-  status <- dat$attr$status
+  status <- get_attr(dat, "status")
+  active <- get_attr(dat, "active")
+  tgl <- get_control(dat, "tergmLite")
```
Comment (smjenness) [inline R/net.mod.infection.R:318]: Use the same name when you pull the object (i.e., tergmLite instead of tgl)

File: R/net.mod.infection.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (smjenness) [reply R/net.mod.infection.R:127]: This can now be updated with #384 merged in

File: R/net.mod.init.R [hunk 9]
```diff
@@ -136,26 +136,39 @@ initialize.net <- function(x, param, init, control, s) {
...
+
   # Variables ---------------------------------------------------------------
-  i.num <- dat$init$i.num
-  i.num.g2 <- dat$init$i.num.g2
-  r.num <- dat$init$r.num
-  r.num.g2 <- dat$init$r.num.g2
+  i.num <- get_init(dat, "i.num", override.null.error = TRUE)
+  if (type  == "SIR"){
+    r.num <- get_init(dat, "r.num", override.null.error = TRUE)
+  }
```
Comment (smjenness) [inline R/net.mod.init.R:155]: shouldn't that NULL be caught with crosscheck?

File: R/net.mod.init.R [hunk 10]
```diff
@@ -136,26 +136,39 @@ initialize.net <- function(x, param, init, control, s) {
...
+      r.num.g2 <- get_init(dat, "r.num.g2")
+    }
   } else {
     group <- rep(1, num)
   }
 
-  type <- dat$control$type
-
+  #TODO: check that this works for tergmLite
+  statOnNw <- "status" %in% dat$temp$nwterms
```
Comment (smjenness) [inline R/net.mod.init.R:171]: You can remove that TODO line now; that is resolved

File: R/net.mod.init.R [hunk 11]
```diff
@@ -200,31 +214,32 @@ init_status.net <- function(dat) {
...
-    if (!is.null(dat$init$infTime.vector)) {
-      infTime <- dat$init$infTime.vector
+    if (!is.null(infTime.vector)) {
+      infTime <- infTime.vector
     } else {
       # If vital dynamics, infTime is a geometric draw over the duration of infection
-      if (dat$param$vital == TRUE && dat$param$di.rate > 0) {
-        if (dat$control$type == "SI") {
-          infTime[idsInf] <- -rgeom(n = length(idsInf), prob = dat$param$di.rate) + 2
+      if (vital && di.rate > 0) {
```
Comment (smjenness) [inline R/net.mod.init.R:223]: make this check on vital explicitly logical (i.e., vital == TRUE)

File: R/net.mod.prevalence.R [hunk 12]
```diff
@@ -122,45 +122,46 @@ prevalence.net <- function(dat, at) {
...
     if (eb == TRUE) {
       for (i in 1:length(ebun)) {
         dat$epi[[paste0("i.num", ebun[i])]] <- sum(status == "i" &
                                                      group == 1 &
                                                      get(ebn) == ebv[i])
       }
     }
-    if (dat$control$type == "SIR") {
-      dat$epi$r.num <- sum(status == "r" & group == 1)
+    if (get_control(dat, "type") == "SIR") {
```
Comment (smjenness) [inline R/net.mod.prevalence.R:163]: pull type at the top

File: R/net.mod.simnet.R [hunk 13]
```diff
@@ -90,15 +96,15 @@ resim_nets <- function(dat, at) {
                            control = dat$control$set.control.stergm))
 
       # Update nwstats df
-      if (dat$control$save.nwstats == TRUE) {
+      if (save.nwstats) {
```
Comment (smjenness) [inline R/net.mod.simnet.R:99]: same comment as above, make the logic check explicit

File: R/net.mod.vital.R [hunk 14]
```diff
@@ -14,71 +14,80 @@
 departures.net <- function(dat, at) {
 
   # Conditions --------------------------------------------------------------
-  if (dat$param$vital == FALSE) {
+  if (!get_param(dat, "vital")) {
```
Comment (smjenness) [inline R/net.mod.vital.R:17]: same issue, explicit logic

File: R/net.utils.R (hunk 6 above)
Comment (smjenness) [reply R/net.utils.R:722]: Yup, revert this

File: tests/testthat/test-net-long.R [hunk 15]
```diff
@@ -170,7 +170,7 @@ test_that("edges models", {
 
   ## "SIR, 1M, CL: 2 sim, set status.vector"
   param <- param.net(inf.prob = 0.5, rec.rate = 0.1)
-  init <- init.net(status.vector = rep(c("s", "i"), each = 50))
+  init <- init.net(status.vector = rep(c("s", "i"), each = 50), r.num = 0)
```
Comment (smjenness) [inline tests/testthat/test-net-long.R:173]: No, it gets passed as a single vector. You can't combine the two approaches like this.

File: R/net.mod.vital.R (hunk 14 above)
Comment (andsv2) [reply R/net.mod.vital.R]: Updated with req. change.

File: tests/testthat/test-net-long.R (hunk 15 above)
Comment (andsv2) [reply tests/testthat/test-net-long.R]: Updated init.net to fix this.

File: R/net.mod.prevalence.R (hunk 12 above)
Comment (andsv2) [reply R/net.mod.prevalence.R]: Formatting change updated.

File: R/net.mod.init.R (hunk 10 above)
Comment (andsv2) [reply R/net.mod.init.R:170]: Removed.

File: R/net.mod.init.R (hunk 9 above)
Comment (andsv2) [reply R/net.mod.init.R:156]: Subsequent change to status.vector to address this reverted. Updated and fixed.

File: R/net.mod.infection.R (hunk 8 above)
Comment (andsv2) [reply R/net.mod.infection.R:321]: Updated and fixed.

File: R/net.utils.R (hunk 6 above)
Comment (andsv2) [reply R/net.utils.R]: Reverted.

File: R/net.mod.nwupdate.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (andsv2) [reply R/net.mod.nwupdate.R:31]: Updated with proper syntax.

File: R/net.mod.infection.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (andsv2) [reply R/net.mod.infection.R:128]: Updated all set_epi_at -> set_epi.

File: R/net.mod.infection.R [hunk 16]
```diff
@@ -115,9 +121,10 @@ infection.net <- function(dat, at) {
 
   ## Save incidence vector
   if (at == 2) {
-    dat$epi$si.flow <- c(0, nInf)
+    dat <- set_epi(dat, "si.flow", 1, 0)
+    dat <- set_epi(dat, "si.flow", 2, nInf)
   } else {
-    dat$epi$si.flow[at] <- nInf
+    dat <- set_epi(dat, "si.flow", at, nInf)
```
Comment (smjenness) [inline R/net.mod.infection.R:127]: No, just this line

File: R/net.mod.infection.R [hunk 17]
```diff
@@ -254,12 +268,13 @@ infection.2g.net <- function(dat, at) {
...
-    dat$epi$si.flow.g2 <- c(0, nInfG2)
-
+    dat <- set_epi(dat, "si.flow", 1, 0)
+    dat <- set_epi(dat, "si.flow", 2, nInf)
+    dat <- set_epi(dat, "si.flow.g2", 1, 0)
+    dat <- set_epi(dat, "si.flow.g2", 2, nInfG2)
   } else {
-    dat$epi$si.flow[at] <- nInf
-    dat$epi$si.flow.g2[at] <- nInfG2
+    dat <- set_epi(dat, "si.flow", at, nInf)
```
Comment (smjenness) [inline R/net.mod.infection.R:276]: same issue

File: R/net.mod.init.R [hunk 18]
```diff
@@ -48,15 +48,15 @@ initialize.net <- function(x, param, init, control, s) {
...
     dat$nwparam <- list(x[-which(names(x) == "fit")])
     groups <- length(unique(get.vertex.attribute(nw, "group")))
-    dat$param$groups <- groups
+    dat <- set_param(dat, "groups", groups)
 
     # Nodal Attributes --------------------------------------------------------
 
     # Standard attributes
     num <- network.size(nw)
     dat$attr$active <- rep(1, num)
```
Comment (smjenness) [inline R/net.mod.init.R:57]: why not use set_attr for active?

File: R/net.mod.init.R [hunk 19]
```diff
@@ -136,26 +136,38 @@ initialize.net <- function(x, param, init, control, s) {
...
+  type <- get_control(dat, "type", override.null.error = TRUE)
+  nsteps <- get_control(dat, "nsteps")
+  tergmLite <- get_control(dat, "tergmLite")
+  vital <- get_param(dat, "vital")
+  groups <- get_param(dat, "groups")
+  status.vector <- get_init(dat, "status.vector", override.null.error = TRUE)
+  if (type %in% c("SIS", "SIR")){
+    rec.rate <- get_param(dat, "rec.rate")
+  }
+  if (vital) {
```
Comment (smjenness) [inline R/net.mod.init.R:148]: explicit logic written

File: R/net.mod.init.R [hunk 20]
```diff
@@ -176,22 +188,23 @@ init_status.net <- function(dat) {
         }
       }
     }
-    dat$attr$status <- status
+    dat <- set_attr(dat, "status", status)
   } else {
-    status <- dat$attr$status
+    #TODO: Where is status set before this? NULL if statonNW = TRUE
```
Comment (smjenness) [inline R/net.mod.init.R:193]: Fix, remove?

File: R/net.mod.nwupdate.R [hunk 21]
```diff
@@ -13,86 +13,95 @@
...
+  vital <- get_param(dat, "vital")
+  tL <- get_control(dat, "tergmLite")
+  status <- get_attr(dat, "status")
+  infTime <- get_attr(dat, "infTime")
+  active <- get_attr(dat, "active")
+  exitTime <- get_attr(dat, "exitTime")
+
+  ## Vital Dynamics
+
+  if (vital) {
```
Comment (smjenness) [inline R/net.mod.nwupdate.R:26]: explicit logic

File: R/net.mod.nwupdate.R [hunk 22]
```diff
@@ -13,86 +13,95 @@
...
+      if (tL == FALSE) {
+        dat$nw[[1]] <- add.vertices(dat$nw[[1]], nv = sum(nArrivals))
+        dat$nw[[1]] <- activate.vertices(dat$nw[[1]], onset = at, terminus = Inf, v = newNodes)
+        dat <- copy_datattr_to_nwattr(dat)
+        dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], prefix = "testatus",
+                                                 value = status[newNodes],
+                                                 onset = at, terminus = Inf,
+                                                 v = newNodes)
+      }
+      if (tL == TRUE) {
```
Comment (smjenness) [inline R/net.mod.nwupdate.R:57]: use tergmLite, instead of tL shorthand

File: R/net.mod.vital.R [hunk 23]
```diff
@@ -14,71 +14,83 @@
...
 
   # Output ------------------------------------------------------------------
 
+  dat <- set_attr(dat, "active", active)
+  dat <- set_attr(dat, "exitTime", exitTime)
+
   if (at == 2) {
-    dat$epi$ds.flow <- c(0, nDepartures.sus)
-    dat$epi$di.flow <- c(0, nDepartures.inf)
+    dat <- set_epi(dat, "di.flow", 1, 0)
```
Comment (smjenness) [inline R/net.mod.vital.R:81]: same comment as above

File: R/net.mod.vital.R [hunk 24]
```diff
@@ -151,92 +165,105 @@ arrivals.net <- function(dat, at) {
...
+  type <- get_control(dat, "type")
+
+  active <- get_attr(dat, "active")
+  status <- get_attr(dat, "status")
+  exitTime <- get_attr(dat, "exitTime")
+  group <- get_attr(dat, "group")
+  rates.sus <- c(get_param(dat, "ds.rate"), get_param(dat, "ds.rate.g2"))
+  rates.inf <- c(get_param(dat, "di.rate"), get_param(dat, "di.rate.g2"))
+  if (type == "SIR")
+  rates.rec <- c(get_param(dat, "dr.rate"), get_param(dat, "dr.rate.g2"))
```
Comment (smjenness) [inline R/net.mod.vital.R:183]: needs curly brackets

File: R/net.mod.vital.R [hunk 25]
```diff
@@ -151,92 +165,105 @@ arrivals.net <- function(dat, at) {
...
       }
     }
   }
 
   # Output ------------------------------------------------------------------
+  dat <- set_attr(dat, "active", active)
   if (at == 2) {
-    dat$epi$ds.flow <- c(0, nDepartures.sus)
-    dat$epi$di.flow <- c(0, nDepartures.inf)
+    dat <- set_epi(dat, "di.flow", 1, 0)
```
Comment (smjenness) [inline R/net.mod.vital.R:241]: same comment as above

File: R/net.mod.vital.R [hunk 26]
```diff
@@ -288,25 +316,26 @@ arrivals.2g.net <- function(dat, at) {
...
+      dat <- append_attr(dat, "entrTime", at, totArr)
+      dat <- append_attr(dat, "exitTime", NA, totArr)
     }
   }
 
   # Output ------------------------------------------------------------------
   if (at == 2) {
-    dat$epi$a.flow <- c(0, nArrivals)
-    dat$epi$a.flow.g2 <- c(0, nArrivalsG2)
+    dat <- set_epi(dat, "a.flow", 1, 0)
```
Comment (smjenness) [inline R/net.mod.vital.R:332]: same comment as above

File: R/net.mod.init.R (hunk 18 above)
Comment (andsv2) [reply R/net.mod.init.R:57]: @smjenness set_attr is based on length of active. Initial setting of active has to be using the initial method, unless we remove the attr length check.

File: R/net.mod.infection.R (hunk 16 above)
Comment (andsv2) [reply R/net.mod.infection.R:127]: @smjenness for this, are we wanting epi trackers at step 1 to record NA instead of 0?

Comment (smjenness) [reply R/net.mod.infection.R:127]: yes

File: R/net.mod.init.R (hunk 18 above)
Comment (smjenness) [reply R/net.mod.init.R:57]: ah, right. let's add an `override.length.check` arg to set_attr that defaults to FALSE, but would be TRUE in this case.

Comment (AdrienLeGuillou) [reply R/net.mod.init.R:57]: on it

File: R/net.mod.vital.R (hunk 24 above)
Comment (andsv2) [reply R/net.mod.vital.R]: Fixed.

File: R/net.mod.init.R (hunk 20 above)
Comment (andsv2) [reply R/net.mod.init.R]: Updated. However: looking through the code it looks like if initial conditions are passed through init.net and status isn't in the formation model (but is in on the network) then status (on the network) is ignored? Is this correct? Should the status on the network only be pulled (eg: through copy_nwattr_to_datarr) if status is in the formation model AND no initial conditions are passed?

Comment (smjenness) [reply R/net.mod.init.R]: I think status passed on the network, whether in the formation model or not, should override  any initial conditions passed through init.net. ok?

File: R/net.mod.nwupdate.R (hunk 21 above)
Comment (andsv2) [reply R/net.mod.nwupdate.R]: Fixed.

File: R/net.mod.vital.R (hunk 23 above)
Comment (andsv2) [reply R/net.mod.vital.R]: Fixed.

File: R/net.mod.vital.R (hunk 25 above)
Comment (andsv2) [reply R/net.mod.vital.R]: Fixed.

File: R/net.mod.nwupdate.R (hunk 22 above)
Comment (andsv2) [reply R/net.mod.nwupdate.R]: All tL logic changed to tergmLite.

File: R/net.mod.init.R (hunk 19 above)
Comment (andsv2) [reply R/net.mod.init.R]: Fixed.

File: R/net.mod.init.R (hunk 18 above)
Comment (andsv2) [reply R/net.mod.init.R:57]: Updated by Adrien.

File: R/net.mod.infection.R (hunk 17 above)
Comment (andsv2) [reply R/net.mod.infection.R]: Syntax fixed.

File: R/net.mod.infection.R (hunk 16 above)
Comment (andsv2) [reply R/net.mod.infection.R]: Syntax fixed.

File: R/net.mod.infection.R (hunk 1 above)
Comment (andsv2) [reply R/net.mod.infection.R]: Overall workflow of get/set functions updated throughout modules.

File: R/net.mod.vital.R (hunk 26 above)
Comment (andsv2) [reply R/net.mod.vital.R]: Syntax fixed.

File: R/net.mod.init.R (hunk 20 above)
Comment (andsv2) [reply R/net.mod.init.R]: statOnNw check has now been updated to look for status on vertex attr. and not whether in the model or not.
